### PR TITLE
Improve social authentication

### PR DESCRIPTION
**Pull request by Marishka17**
### How has this been tested?

In another [PR](https://github.com/opencv/cvat/pull/5444)
### Checklist

- [x] I submit my changes into the `develop` branch
~~- [ ] I have added a description of my changes into [CHANGELOG](https://github.com/cvat-ai/cvat/blob/develop/CHANGELOG.md) file~~
~~- [ ] I have updated the [documentation](
  https://github.com/cvat-ai/cvat/blob/develop/README.md#documentation) accordingly~~
- [x] I have added tests to cover my changes
~~- [ ] I have linked related issues ([read github docs](
  https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword))~~
- [ ] I have increased versions of npm packages if it is necessary ([cvat-canvas](https://github.com/cvat-ai/cvat/tree/develop/cvat-canvas#versioning),
[cvat-core](https://github.com/cvat-ai/cvat/tree/develop/cvat-core#versioning), [cvat-data](https://github.com/cvat-ai/cvat/tree/develop/cvat-data#versioning) and [cvat-ui](https://github.com/cvat-ai/cvat/tree/develop/cvat-ui#versioning))

### License

- [x] I submit _my code changes_ under the same [MIT License](
  https://github.com/cvat-ai/cvat/blob/develop/LICENSE) that covers the project.
  Feel free to contact the maintainers if that's a concern.

### CODE REVIEW

File: cvat-ui/src/components/login-with-social-app/login-with-social-app.tsx [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+    useEffect(() => {
+        if (provider && code && state) {
+            const req: AxiosRequestConfig = {
+                method: 'POST',
+                data: {
+                    code,
+                },
+            };
+            cvat.server.request(
+                `${cvat.config.backendAPI}/auth/${provider}/login/token?state=${state}`, req,
```
Comment (nmanovic) [inline cvat-ui/src/components/login-with-social-app/login-with-social-app.tsx:32]: @Marishka17 , should it be a part of cvat-core?

File: cvat/apps/iam/views.py [hunk 2]
```diff
@@ -215,10 +217,17 @@ def get(self, request):
...
-        return super().dispatch(request, *args, **kwargs)
+        if (auth_error := request.GET.get('error', None)):
+            if auth_error == self.adapter.login_cancelled_error:
+                return HttpResponseRedirect(settings.SOCIALACCOUNT_CALLBACK_CANCELLED_URL)
+            else: # unknown error
+                raise ValidationError(auth_error)
+
+        code = request.GET.get('code')
+        state = request.GET.get('state')
+        assert code and state, 'Parameters code and state not found in request'
```
Comment (nmanovic) [inline cvat/apps/iam/views.py:228]: It should not an assertion. Our users can use our API incorrectly. Need to return BadRequest error with corresponding error.

**Review by nmanovic — CHANGES_REQUESTED** [review]
@Marishka17 , it looks good from my perspective at least. Need to slightly change the code. Also I expect to see all these auth methods in our swagger. What is a reason to hide them?

File: cvat-ui/src/components/login-with-social-app/login-with-social-app.tsx [hunk 3]
```diff
@@ -0,0 +1,53 @@
+// Copyright (C) 2022 CVAT.ai Corporation
+//
+// SPDX-License-Identifier: MIT
+
+import React, { useEffect } from 'react';
+import { useLocation, useHistory } from 'react-router';
+import notification from 'antd/lib/notification';
+
+import { getCore } from '../../cvat-core-wrapper';
```
Comment (bsekachev) [inline cvat-ui/src/components/login-with-social-app/login-with-social-app.tsx:9]:
```suggestion
import { getCore } from 'cvat-core-wrapper';
```

File: cvat-ui/src/components/login-with-social-app/login-with-social-app.tsx (hunk 1 above)
Comment (bsekachev) [reply cvat-ui/src/components/login-with-social-app/login-with-social-app.tsx:32]: Do you mean dedicated API method like ``loginSocial``?

File: cvat-ui/src/components/login-with-social-app/login-with-social-app.tsx [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+                    }
+                    history.push('/auth/login');
+                    notification.error({
+                        message: 'Could not log in with social account',
+                        description: 'Go to developer console',
+                    });
+                    return Promise.reject(exception);
+                });
+        }
+    }, [provider, code, state]);
```
Comment (bsekachev) [inline cvat-ui/src/components/login-with-social-app/login-with-social-app.tsx:50]: Probably argument list should be empty and for code clearness we should extract ``provider, code, state`` inside the ``useEffect`` body.

File: cvat-ui/src/components/login-with-social-app/login-with-social-app.tsx [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+
+    return <></>;
```
Comment (bsekachev) [inline cvat-ui/src/components/login-with-social-app/login-with-social-app.tsx:52]: I would suggest return spinner here, look at logout component, for example

**Review by nmanovic — APPROVED** [review]

**Pull request merged**